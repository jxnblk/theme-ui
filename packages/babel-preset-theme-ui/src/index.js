module.exports = () => ({
  plugins: [
    [
      '@wordpress/babel-plugin-import-jsx-pragma',
      {
        source: 'theme-ui',
        scopeVariable: 'jsx',
        isDefault: false,
      },
    ],
    [
      '@babel/plugin-transform-react-jsx',
      {
        pragma: 'jsx',
      },
    ],
  ],
})