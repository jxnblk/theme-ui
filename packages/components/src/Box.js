import styled from '@emotion/styled'
import { css, get } from '@theme-ui/css'
import { merge } from '@theme-ui/core'
import { createShouldForwardProp } from '@styled-system/should-forward-prop'
import space from '@styled-system/space'
import color from '@styled-system/color'

const shouldForwardProp = createShouldForwardProp([
  ...space.propNames,
  ...color.propNames,
])

const sx = props => css(props.sx)(props.theme)
const base = props => css(props.__css)(props.theme)
const variant = ({ theme, variant, __themeKey = 'variants' }) =>
  css(get(theme, __themeKey + '.' + variant, get(theme, variant)))
const variants = ({ theme, variants = [], __themeKey = 'variants' }) =>
  css(merge(variants.map(v => get(theme, __themeKey + '.' + v, get(theme, v)))))

export const Box = styled('div', {
  shouldForwardProp,
})(
  {
    boxSizing: 'border-box',
    margin: 0,
    minWidth: 0,
  },
  base,
  variants,
  variant,
  space,
  color,
  sx,
  props => props.css
)

export default Box
