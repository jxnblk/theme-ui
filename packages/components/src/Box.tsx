import {
  ArrayInterpolation,
  CSSObject,
  Interpolation,
  useTheme,
} from '@emotion/react'
import React, { forwardRef } from 'react'
import { css, get, ThemeUIStyleObject } from '@theme-ui/css'
import type { Assign, ForwardRef } from './types'
import type { __ThemeUIComponentsInternalProps } from './util'

export interface BoxOwnProps {
  as?: React.ElementType
  variant?: string
  css?: Interpolation<any>
  sx?: ThemeUIStyleObject
}

export interface BoxProps
  extends Omit<
    Assign<React.ComponentPropsWithRef<'div'>, BoxOwnProps>,
    'ref'
  > {}

/**
 * Use the Box component as a layout primitive to add margin, padding, and colors to content.
 * @see https://theme-ui.com/components/box
 */
export const Box: ForwardRef<any, BoxProps> = forwardRef<any, BoxProps>(
  function Box(props, ref) {
    const theme = useTheme()

    const {
      __themeKey = 'variants',
      __css,
      variant,
      css: cssProp,
      sx,
      as: Component = 'div',
      ...rest
    } = props as BoxProps & __ThemeUIComponentsInternalProps

    const baseStyles: CSSObject = {
      boxSizing: 'border-box',
      margin: 0,
      minWidth: 0,
    }

    const __cssStyles = css(__css)(theme)

    const variantInTheme =
      get(theme, `${__themeKey}.${variant}`) || get(theme, variant)
    const variantStyles = variantInTheme && css(variantInTheme)(theme)

    const sxPropStyles = css(sx)(theme)

    const style: ArrayInterpolation<unknown> = [
      baseStyles,
      __cssStyles,
      variantStyles,
      sxPropStyles,

      cssProp,
    ]

    return <Component ref={ref} css={style} {...rest} />
  }
)
