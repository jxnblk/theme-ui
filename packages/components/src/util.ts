import type { ThemeUICSSObject, ThemeUICSSProperties } from '@theme-ui/css'

/** @internal */
export function __internalProps(props: __ThemeUIComponentsInternalProps) {
  return props as {}
}

/**
 * @internal Props used by Theme UI Components not intended for user code.
 */
export interface __ThemeUIComponentsInternalProps {
  __css?: ThemeUICSSObject
  __themeKey?: string
}
