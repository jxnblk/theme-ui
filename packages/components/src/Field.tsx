import React from 'react'

import { Box } from './Box'
import { Label } from './Label'
import { Input, InputProps } from './Input'
import type { ThemeUICSSObject } from '@theme-ui/css'

export interface FieldOwnProps {
  /**
   * Text for Label component
   */
  label?: string
  /**
   * Used for the for, id, and name attributes
   */
  name?: string
  containerSx?: ThemeUICSSObject
}

export type FieldProps<T extends React.ElementType> = FieldOwnProps &
  Omit<React.ComponentPropsWithRef<T>, 'as' | keyof FieldOwnProps> & {
    /**
     * form control to render, default Input
     */
    as?: T
  }

export interface Field {
  <T extends React.ElementType = React.ComponentType<InputProps>>(
    props: FieldProps<T>
  ): JSX.Element
}

export const Field = React.forwardRef(function Field<
  T extends React.ElementType = React.ComponentType<InputProps>
>(
  {
    // if somebody specifies the generic parameter without passing `as` prop, they get Input anyway
    as: Control = Input as any as T,
    label,
    id,
    name,
    containerSx,
    ...rest
  }: FieldProps<T>,
  ref: React.ForwardedRef<unknown>
) {
  const fieldIdentifier = id || name

  const controlProps = {
    ref,
    name,
    id: fieldIdentifier,
    ...rest,
  } as React.ComponentPropsWithRef<T>

  return (
    <Box sx={containerSx}>
      <Label htmlFor={fieldIdentifier}>{label}</Label>
      <Control {...controlProps} />
    </Box>
  )
}) as Field // Field is generic bcs of `as` prop, so we can't just use types from forwardRef
