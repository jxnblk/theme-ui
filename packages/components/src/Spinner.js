import React from 'react'
import { keyframes } from '@emotion/core'
import Box from './Box'

const range = n => {
  const v = []
  for (let i = 0; i < n; i++) {
    v.push(i)
  }
  return v
}

const spin = keyframes({
  from: {
    transform: 'rotate(0deg)',
  },
  to: {
    transform: 'rotate(360deg)',
  },
})

const roller = keyframes({
  '0%, 20%, 80%, 100%': {
    transform: 'scale(1)',
  },
  '50%': {
    transform: 'scale(1.5)',
  },
})

const facebook = keyframes({
  '0%': {
    top: '8px',
    height: '48px',
  },
  '50%, 100%': {
    top: '24px',
    height: '24px',
  },
})

const animations = { spin, roller, facebook }

export const Spinner = props => {
  const {
    animation = 'spin',
    duration = 600,
    width = 48,
    height = 48,
    items = 1,
    delay,
    itemX = () => 0,
    itemY = () => 0,
    itemWidth,
    itemHeight,
  } = props

  return (
    <Box
      role="img"
      {...props}
      __themeKey="spinners"
      sx={{
        position: 'relative',
        display: 'inline-block',
        width: `${width}px`,
        height: `${height}px`,
      }}>
      {range(items).map(index => {
        const cfg = { ...props, index, width, height }
        return (
          <Box
            key={index}
            sx={{
              position: 'absolute',
              display: 'block',
              width: `${itemWidth || width}px`,
              height: `${itemHeight || height}px`,
              animationIterationCount: 'infinite',
              animationName: animations[animation].toString(),
              animationDuration: `${duration}ms`,
              animationDelay: `${(delay || 0) * index}ms`,
              top: `${itemY(cfg)}px`,
              left: `${itemX(cfg)}px`,
            }}
          />
        )
      })}
    </Box>
  )
}

export const DualringSpinner = ({ duration }) => (
  <Spinner
    animation="spin"
    duration={duration}
    delay={-150}
    items={4}
    __css={{
      borderRadius: '50%',
      borderWidth: 6,
      borderStyle: 'solid',
      borderColor: 'rgba(51, 51, 238, 0.1)',
      '& > div': {
        content: '" "',
        borderRadius: '50%',
        left: `-6px`,
        top: `-6px`,
        borderWidth: 6,
        borderStyle: 'solid',
        borderColor: `#3333ee transparent transparent transparent`,
        animationTimingFunction: 'cubic-bezier(0.5, 0, 0.5, 1)',
      },
    }}
  />
)

export const FacebookSpinner = ({ duration, itemWidth, itemHeight }) => (
  <Spinner
    duration={duration}
    items={3}
    itemWidth={itemWidth}
    itemHeight={itemHeight}
    animation="facebook"
    delay={100}
    itemX={({ index, itemWidth }) => (8 + itemWidth) * index}
    __css={{
      '& > div': {
        display: 'inline-block',
        background: '#3333ee',
        animationTimingFunction: 'cubic-bezier(0, 0.5, 0.5, 1)',
      },
    }}
  />
)

export const RollerSpinner = ({ duration, items, itemWidth, itemHeight }) => (
  <Spinner
    duration={duration}
    items={items}
    animation="roller"
    delay={100}
    itemX={({ index, items, width }) => {
      const r = width / 2
      const angle = (index / items) * Math.PI * 2
      return r + r * Math.cos(angle)
    }}
    itemY={({ index, items, height }) => {
      const r = height / 2
      const angle = (index / items) * Math.PI * 2
      return r + r * Math.sin(angle)
    }}
    itemWidth={itemWidth}
    itemHeight={itemHeight}
    __css={{
      '& > div': {
        background: '#3333ee',
        borderRadius: '50%',
        animationTimingFunction: 'linear',
      },
    }}
  />
)
