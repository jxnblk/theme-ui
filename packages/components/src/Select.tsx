import React from 'react'

import { get, ThemeUICSSObject } from '@theme-ui/css'

import { Box, BoxOwnProps, BoxProps } from './Box'
import { SVG, SVGProps } from './SVG'
import { __internalProps } from './util'
import type { Assign, ForwardRef } from './types'

const DownArrow = (props: SVGProps) => (
  <SVG {...props}>
    <path d="M7 10l5 5 5-5z" />
  </SVG>
)

export interface SelectProps
  extends Assign<React.ComponentPropsWithRef<'select'>, BoxOwnProps> {
  arrow?: React.ReactElement
  containerSx?: ThemeUICSSObject
}

/**
 * Select variants can be defined in `theme.forms`
 * and the component uses the `theme.forms.select` variant by default.
 * @see https://theme-ui.com/components/select/
 */
export const Select: ForwardRef<HTMLSelectElement, SelectProps> =
  React.forwardRef(function Select({ arrow, containerSx, ...props }, ref) {
    const __css: ThemeUICSSObject = {
      display: 'block',
      width: '100%',
      p: 2,
      paddingRight: 4,
      appearance: 'none',
      fontSize: 'inherit',
      lineHeight: 'inherit',
      border: '1px solid',
      borderRadius: 4,
      color: 'inherit',
      backgroundColor: (theme) => get(theme, 'colors.background', null),
    }

    return (
      <Box
        sx={{
          display: 'flex',
          ...containerSx,
        }}
      >
        <Box
          ref={ref}
          as="select"
          variant="select"
          {...(props as BoxProps)}
          {...__internalProps({ __themeKey: 'forms', __css })}
        />
        {arrow || (
          <DownArrow
            sx={{
              ml: -28,
              alignSelf: 'center',
              pointerEvents: 'none',
            }}
          />
        )}
      </Box>
    )
  })
