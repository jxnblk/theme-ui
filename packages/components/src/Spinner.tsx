import React from 'react'
import type { ThemeUICSSObject } from '@theme-ui/css'

import { Box, BoxOwnProps, BoxProps } from './Box'
import type { ForwardRef } from './types'
import { __internalProps } from './util'

export interface SpinnerProps
  extends Omit<
      React.ComponentPropsWithRef<'svg'>,
      'opacity' | 'color' | 'css' | 'sx' | 'strokeWidth'
    >,
    BoxOwnProps {
  size?: number
  strokeWidth?: number
  title?: string
  duration?: number
}

export const Spinner: ForwardRef<SVGSVGElement, SpinnerProps> =
  React.forwardRef(function Spinner(
    {
      size = 48,
      strokeWidth = 4,
      max = 1,
      title = 'Loadingâ€¦',
      duration = 750,
      ...props
    },
    ref
  ) {
    const __css: ThemeUICSSObject = {
      color: 'primary',
      overflow: 'visible',
    }

    const svgProps = {
      strokeWidth,

      viewBox: `0 0 ${size} ${size}`,
      width: size,
      height: size,

      fill: 'none',
      stroke: 'currentColor',
      role: 'img',
    }

    const radius = size / 2 - strokeWidth
    const center = radius + strokeWidth
    const circleProps = {
      strokeWidth,
      r: radius,
      cx: radius + strokeWidth,
      cy: radius + strokeWidth,
      fill: 'none',
    }

    return (
      <Box
        ref={ref}
        as="svg"
        {...(svgProps as {})}
        {...(props as BoxProps)}
        {...__internalProps({ __css })}
      >
        <title>{title}</title>
        <circle {...circleProps} opacity={1 / 8} />
        <circle {...circleProps} strokeDasharray="20 180">
          <animateTransform
            attributeName="transform"
            attributeType="XML"
            type="rotate"
            from={`0 ${center} ${center}`}
            to={`360 ${center} ${center}`}
            dur={`${duration}ms`}
            repeatCount="indefinite"
          />
        </circle>
      </Box>
    )
  })
