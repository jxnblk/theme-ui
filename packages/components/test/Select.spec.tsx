/**
 * @jest-environment jsdom
 */

import { ThemeProvider } from '@theme-ui/core'
import { renderJSON } from '@theme-ui/test-utils'
import React from 'react'

import { Select, SelectProps } from '../src'

import { theme } from './__test-utils__'

describe('Select', () => {
  test('renders', () => {
    const json = renderJSON(
      <ThemeProvider theme={theme}>
        <Select />
      </ThemeProvider>
    )
    expect(json).toMatchSnapshot()
  })

  test('renders with sx prop', () => {
    const json = renderJSON(
      <ThemeProvider theme={theme}>
        <Select sx={{ ml: 3 }} value="hello" />
      </ThemeProvider>
    )
    expect(json).toMatchSnapshot()
  })

  test('renders with custom icon', () => {
    const json = renderJSON(
      <ThemeProvider theme={theme}>
        <Select
          arrow={
            <svg>
              <path d="M7.41 7.84l4.59 4.58 4.59-4.58 1.41 1.41-6 6-6-6z" />
            </svg>
          }
        />
      </ThemeProvider>
    )
    expect(json).toMatchSnapshot()
  })

  test('renders with background-color', () => {
    const json = renderJSON(
      <ThemeProvider theme={{ colors: { background: 'blueviolet' } }}>
        <Select />
      </ThemeProvider>
    )!

    expect(json.children?.[0]).toHaveStyleRule('background-color', 'blueviolet')
  })
})
