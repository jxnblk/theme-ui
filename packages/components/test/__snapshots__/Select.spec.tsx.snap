// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  padding-right: 32px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: -28px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  pointer-events: none;
}

<div
  className="emotion-0"
>
  <select
    className="emotion-1"
  />
  <svg
    className="emotion-2"
    fill="currentcolor"
    height={24}
    viewBox="0 0 24 24"
    width={24}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
  </svg>
</div>
`;

exports[`Select renders with custom icon 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  padding-right: 32px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
}

<div
  className="emotion-0"
>
  <select
    className="emotion-1"
  />
  <svg>
    <path
      d="M7.41 7.84l4.59 4.58 4.59-4.58 1.41 1.41-6 6-6-6z"
    />
  </svg>
</div>
`;

exports[`Select renders with sx prop 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  padding-right: 32px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  margin-left: 16px;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: -28px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  pointer-events: none;
}

<div
  className="emotion-0"
>
  <select
    className="emotion-1"
    value="hello"
  />
  <svg
    className="emotion-2"
    fill="currentcolor"
    height={24}
    viewBox="0 0 24 24"
    width={24}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
  </svg>
</div>
`;
